bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java

```diff
@@ -1724,63 +1797,85 @@ private boolean verifyCookie(Cookie oldCookie, File dir) throws IOException {
+                    }
+                } else {
+                    LOG.info("Listing local Cookie Files of Bookie: {}", bookieAddress);
+                    for (File verFile : allVersionFiles) {
+                        LOG.info(verFile.getCanonicalPath());
+                    }
+                }
+            } else {
+                LOG.info("No local cookies for Bookie: {}", bookieAddress);
+            }
 
-        @Override
-        int runCmd(CommandLine cmdLine) {
-            ServerConfiguration conf = new ServerConfiguration(bkConf);
             try (RegistrationManager rm = new ZKRegistrationManager()) {
+                Versioned<Cookie> cookie = null;
                 try {
-                    rm.initialize(bkConf, () -> {
-                    }, NullStatsLogger.INSTANCE);
-                } catch (BookieException e) {
-                    LOG.error("Exception while establishing zookeeper connection.", e);
-                    return -1;
+                    rm.initialize(bkConf, () -> { }, NullStatsLogger.INSTANCE);
+                    cookie = Cookie.readFromRegistrationManager(rm, bookieAddress);
+                } catch (CookieNotFoundException nne) {
+                    LOG.info("No cookie for {} in ZooKeeper", bookieAddress);
+                    return 0;
                 }
 
-                List<File> allLedgerDirs = Lists.newArrayList();
-                allLedgerDirs.addAll(Arrays.asList(ledgerDirectories));
-                if (indexDirectories != ledgerDirectories) {
-                    allLedgerDirs.addAll(Arrays.asList(indexDirectories));
-                }
+                if (delete) {
+                    boolean confirm = force;
+                    if (!confirm) {
+                        confirm = IOUtils.confirmPrompt("Are you sure you want to delete Cookies from Zookeeper?");
+                    }
 
-                try {
-                    conf.setAllowStorageExpansion(true);
-                    Bookie.checkEnvironmentWithStorageExpansion(conf, rm,
-                            Lists.newArrayList(journalDirectories), allLedgerDirs);
-                } catch (BookieException e) {
-                    LOG.error(
-                            "Exception while updating cookie for storage expansion", e);
-                    return -1;
+                    if (confirm) {
+                        cookie.getValue().deleteFromRegistrationManager(rm, bkConf, cookie.getVersion());
+                        LOG.info("Deleted Cookie from Zookeeper for Bookie: {}", bookieAddress);
+                    } else {
+                        LOG.info("Skipping deleting cookie from ZooKeeper for Bookie: {}", bookieAddress);
+                    }
+                } else {
+                    LOG.info("{} bookie's Cookie path in Zookeeper: {} ", bookieAddress,
+                            rm.getCookiePath(bookieAddress.toString()));
```

Comment:
do we really need line 1874-1875? I am not sure if it is worth making a getCookiePath interface in RegistrationManager interface just for this purpose.

Reply:
in this particular command, we are intending to print cookiePath, which depends on RegistrationManager. 

If we don't want to have this method at RegistrationManager interface level, then we can remove it from interface and typecast RegistrationManager instance to ZKRegistrationManager and call method of ZKRegistrationManager.

If you have any other thoughts, I'm open.

Reply:
I would suggest leave this as part of ZKRegistraitonManager for now. Because it is internal to zookeeper implementation.

You don't need to do cast. You can change `try (RegistrationManager rm = new ZKRegistrationManager())` to `try (ZKRegistrationManager rm = new ZKRegistrationManager())`.

Reply:
done